src/globus_cli/commands/ls.py

```diff
@@ -180,29 +180,32 @@ def ls_command(
     ls_params: dict[str, t.Any] = {"show_hidden": int(show_hidden)}
     if path:
         ls_params["path"] = path
-    if filter_val:
-        # this char has special meaning in the LS API's filter clause
-        # can't be part of the pattern (but we don't support globbing across
-        # dir structures anyway)
-        if "/" in filter_val:
-            raise click.UsageError('--filter cannot contain "/"')
-        # format into a simple filter clause which operates on filenames
-        ls_params["filter"] = f"name:{filter_val}"
+
+    # this char has special meaning in the LS API's filter clause
+    # can't be part of the pattern (but we don't support globbing across
+    # dir structures anyway)
+    if filter_val and "/" in filter_val:
+        raise click.UsageError('--filter cannot contain "/"')
 
     # get the `ls` result
     if recursive:
-        # NOTE:
-        # --recursive and --filter have an interplay that some users may find
-        # surprising
-        # if we're asked to change or "improve" the behavior in the future, we
-        # could do so with "type:dir" or "type:file" filters added in, and
-        # potentially work out some viable behavior based on what people want
+
+        # if we are doing filtering we need to pass multiple filter params. The
+        # first allows all directories, as we need them for recursive
+        # expansion. The second then filters name by the filter_val
+        if filter_val:
+            ls_params["filter"] = ["type:dir", f"name:{filter_val}"]
```

Comment:
If I recall our formatting code correctly, we can also leave the `:` formatting to the SDK, like so:
```suggestion
            ls_params["filter"] = [{"type": "dir"}, {"name": filter_val}]
```

Do we like this better? I think it's pretty much the same, but maybe others will find it easier to read or have a stronger opinion.

Reply:
Oh, I like that a bit more, I'm just going to add it